agol-publish/NightStocker.py

```diff
@@ -0,0 +1,518 @@
+             case)
+    '''
+
+    updated_row = None
+
+    client = pygsheets.authorize(service_file=gsheet_auth)
+
+    #: Update stewardship doc
+    sheet = client.open_by_key(gsheet_keys[0])
+    worksheet = sheet[1]  #: Stewardship sheet is second tab
+
+    #: Get all rows so we can work locally before update_values()
+    rows = []
+    for row in worksheet:
+        rows.append(row)
+    #: Row Structure:
+    #: [0 Issue, 1 Authoritative Access From, 2 SGID Data Layer,
+    #: 3 Refresh Cycle (Days), 4 Last Update, 5 Days From Last Refresh,
+    #: 6 Days to Refresh, 7 Description, 8 Data Source, 9 Use Restrictions,
+    #: 10 Website URL, 11 Data Type, 12 PEL Layer, 13 PEL Status,
+    #: 14 Governance/Agreement, 15 PEL Inclusion, 16 Agency Contact Name,
+    #: 17 Agency Contact Email, 18 SGID Coordination, 19 Archival Schedule,
+    #: 20 Endpoint, 21 Tier, 22 Webapp, 23 Notes, 24 Deprecated]
+
+    #: Action info:
+    #: [0 AGOL title, 1 operation, 2 SGID name for stewardship doc, 
+    #: 3 description, 4 source/credit, 5 shape type, 6 endpoint, 7 AGOL item ID]
+
+    updated = False
+
+    for i, row in enumerate(rows):
+        if row[2] == action_info[2]:
+            temp_row = row
+            temp_row[1] = 'AGRC AGOL'
+            temp_row[20] = action_info[6]
+            temp_row[23] = f'AGOL category: {action_info[1]} - {row[23]}'
+            rownum = i+1
+            start = f'A{rownum}'
+            worksheet.update_values(start, [temp_row])
+            updated = True
+            updated_row = rownum
+
+    if not updated:
+        print(f'{action_info[2]} not found in stewardship doc')
+        new_row = []
+        new_row.append('')  #: Leading Note 
+        new_row.append('AGRC AGOL')  #: current source
+        new_row.append(action_info[2])
+        new_row.append('Static')  #: refresh cycle
+        new_row.append('')  #: Last update
+        new_row.append('')  #: Days from last update
+        new_row.append('')  #: Days to refresh
+        new_row.append(sub('<[^<]+?>', '', action_info[3]).strip())  #: Description
+        new_row.append(action_info[4])  #: Data Source
+        new_row.append('')  #:  Use Restrictions
+        new_row.append('')  #:  Website URL
+        new_row.append(action_info[5])  #: Data type
+        new_row.append('')  #: PEL Layer
+        new_row.append('')  #: PEL Status
+        new_row.append('')  #: Governance/Agreement
+        new_row.append('')  #: PEL Inclusion
+        new_row.append('')  #: Agency Contact Name
+        new_row.append('')  #: Agency Contact Email
+        new_row.append('')  #: SGID Coordination
+        new_row.append('')  #: Archival Schedule
+        new_row.append(action_info[6])  #: Endpoint
+        new_row.append('')  #: Tier
+        new_row.append('')  #: Webapp
+        new_row.append(f'Added by NightStocker - AGOL category: {action_info[1]}')  #: Notes
+        new_row.append('')  #: Deprecated
```

Comment:
I think there is a way to use a header row so you don't have to deal with the empties and if the schema changes it wouldn't all break.

Reply:
I guess that would work if I read the whole worksheet in as a data frame, then add a new row to the dataframe using the column titles, then get that row as a list and pass that to `.insert_rows`. Is that what you were thinking, or do you have a different method in mind?

Reply:
i don't think you would need to read the entire worksheet, just the first row. Then you could get the indexes of the fields you needed?

Reply:
I guess I'm not understanding what you're suggesting. For `.insert_rows()` I need a list that is the same length as the spreadsheet columns, and the columns I want to be blank (which as you can see is a lot of them) need to be empty strings. I suppose I could get the header line as a list, then loop through the list changing values to either empty strings or the new values and then use that in `.insert_rows()'. Is this what you mean?

Reply:
I had https://github.com/agrc/gis.utah.gov/blob/master/scripts/datatable.py#L92-L106 in mind. 
If you got the first row, you would know the length so you could use `for index in range()`. If you could get the indexes of the rows you wanted like in the link above, then you could push an empty value to the records you don't care about and based on the index of the fields you do care about, push the correct values.